doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+server-speaks-first protocols like MySQL and SMTP, necessitating the need to disable such
+detection by policy.
+
+In order to detect the connection's session protocol transparently (i.e. without policy),
+the proxy reads from the connection, buffering data until the proxy determines how to
+handle the connection.
+
+#### TLS
+
+If the connection begins with a TLS[[1]](#tls-q1) client handshake, data is read until
+until a full `ClientHello` message is buffered. The proxy determines whether it should
+terminate the TLS connection or treat the connection as an opaque TCP stream. The proxy
+may terminate TLS when a Conduit-specific extension[[2]](#tls-q2) exists in the client's
+hello, or it may terminate TLS if it is configured with a certificate for the
+client-requested name (i.e. for external/ingress).
+
+If the proxy elects to terminate TLS, the client's ALPN[[3]](#tls-q3) advertisement is
+used to determine which protocol should be used. If the ALPN advertisement does not
+include a proxy-supported protocol (i.e. `http` or `h2`), then the proxy handles the
+decrypted stream as opaque TCP[4](#tls-q4).
+
+##### Questions
+
+<a name="tls-q1"></a>
+1. What's the minimum version of the TLS protocol we support? I assume we don't handle SSL
+    3.0, etc. due to ClientHello requirement
+<a name="tls-q2"></a>
+2. How do TLS extensions play into this detection? are they conduit-specific or are we
+   reusing existing public extensions?
+<a name="tls-q3"></a>
+3. Can we reasonably expect clients to support ALPN? Do we have to support detecting
+   protocols for TLS requests that don't advertise a protocol?
+<a name="tls-q4"></a>
+4. It's not immediately clear to me how the proxy can "bridge" two TLS
+   connections--propagating ALPN information could be complex: we might want to issue a
+   `ClientHello` to the destination before issuing a ServerHello to the initiatior. This is
+   mostly germane for ingress.
+
+
+#### HTTP
+
+If the connection begins with an HTTP/1 request-line, the connection is handled as HTTP/1.
+
+If the connection begins with a HTTP/2 Prior-knowledge preface, the connection is handled
+as HTTP/2.
+
+
+## TCP Forwarding
+
+If, after protocol detection (and potentially TLS decryption), the connection's protocol
+is not supported by the proxy, the connection is simply forwarded to the original
+destination.
+
+Policy may require that the forwarded connections are TLS'd by the proxy (i.e. for mutual
+authentication).
+
+
+## HTTP Proxying
+
+### HTTP/1.0
+
+If the request is an HTTP/1.0 request that does not include a `Host` header, the request
+is routed to the connection's original destination address. The request is sent on a
+dedicated one-off connection (in keeping with the HTTP/1.0 specification).
+
+If an HTTP/1.0 request is received and _does_ include a `Host` header, the request is
+routed as if it were HTTP/1.1, though the proxy must be sure to honor proper HTTP/1.0
+semantics to the client peer.
```

Comment:
The proxy needs to use HTTP/1.0 semantics with both the sender and the receiver, because it doesn't know which HTTP versions each side supports.

Reply:
Are you suggesting that the following shouldn't be supported?
```
Client <-- http/1.0 --> Proxy <-- http/1.1 --> Server
```

Linkerd does this today, and it tends to be a big improvement for poorly configured clients. Specifically, if we _can_ maintain long-lived connections to a server, we should do so.

Furthemore, via [mnot](https://www.mnot.net/blog/2011/07/11/what_proxies_must_do.html):

> 0. Advertise HTTP/1.1 Correctly
>
> HTTP/1.1 is the current version spoken on the Internet, and as long as the proxy implements it correctly (in particular, it handles chunked encoding correctly), it should always advertise itself as HTTP/1.1 conformant.
> 
> This means that the top line of requests and responses should always contain HTTP/1.1 as the version identifier, even when talking to something that says it’s HTTP/1.0.
> 
> The reason for this is that HTTP 1.1 defines not only how to talk to 1.1 devices, but also to 1.0 ones. When a HTTP/1.0 message contains a 1.1 mechanism like Cache-Control, its meaning doesn’t change, and should still be respected.
> 
> See [the spec](http://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-14#section-2.5) for more information.

Reply:
I'm personally in favor of the proxy trying to improve HTTP communication where it can, which includes trying to use HTTP/1.1 even if one side claims HTTP/1.0. I don't mean assume 1.1 semantics in a message labeled 1.0, but that we should advertise we know 1.1, and make use of relevant features that are supposed to be forwards-compatible.

I'm a little sympathetic to the case where clients or servers only support 1.0, and are so poorly written that they blow up otherwise, but how often do these implementations exist?

Reply:
> Are you suggesting that the following shouldn't be supported?
> 
>> Client <-- http/1.0 --> Proxy <-- http/1.1 --> Server

In the short term: yes, I'm suggesting we don't bother with that. Instead, I am proposing that we focus on the following, in priority order:

1. Be as transparent as possible by default.
2. Upgrade plaintext to TLS (i.e. add TLS transparently) for any kind of conduit-to-conduit traffic.
3. Upgrade plaintext HTTP/1.1 to HTTP/2 + TLS for conduit-to-conduit connections.

> Linkerd does this today, and it tends to be a big improvement for poorly configured clients. Specifically, if we can maintain long-lived connections to a server, we should do so.

I agree that this is something that we should try to do in the longer term. However, I don't think we should try to do it now for two reasons:
1. It is really important that we demonstrate that the Conduit proxy doesn't break anything when injected in its default configuration. We want to have the mutating webhook admission controller be as generally applicable as possible (ideally, we want to be able to tell people "just inject conduit into everything"), which makes this *really* important. Another way of saying this is that we want injecting conduit to be, first and foremost, a non-event. This means we should try to change as little as possible in an application's traffic by default, and focus on getting the things we do change 100% stable and fast.
2. We expect all conduit-to-conduit traffic to be HTTP/2 + TLS by default, right? If so, speaking HTTP/1.1 outbound is really not too interesting in the short term.
3. Why would a user have an HTTP/1.0 client that does connection management badly in the first place? Almost every HTTP framework and library is doing HTTP/1.1 by default. It's likely because either the server it is connecting to does HTTP poorly or because this is a legacy "don't change anything" system that's basically maintainable. I think we should be conservative in the default configuration, at least initially, in how we handle such systems.

I don't mean to say we'd *never* support such upgrades, but I do think we have such a long pipeline of more urgent things to do that we could/should disregard or at least de-emphasis these trickier and less commonly needed (micro-)features.

Reply:
> I'm personally in favor of the proxy trying to improve HTTP communication where it can, which includes trying to use HTTP/1.1 even if one side claims HTTP/1.0. I don't mean assume 1.1 semantics in a message labeled 1.0, but that we should advertise we know 1.1, and make use of relevant features that are supposed to be forwards-compatible.

In the abstract, I agree. I think, though, that upgrading is than it seems at first. For example, if we translate "HTTP/1.0" in a request from the proxied app to "HTTP/1.1" in our outbound request then we are responsible for translating (roughly) 100% of the differences between HTTP/1.0 and HTTP/1.1. For example, advertising HTTP/1.1 in our request means we support chunked transfer encoding in responses, and we'd have to translate that into a non-chunked response. That means adding the Content-Length, which means calculating the Content-Length, which means buffering the entire response body. But, we, in generally, cannot have buffers of unbounded length in the Conduit proxy. Thus, it seems like we cannot translate HTTP/1.1 responses to HTTP/1.0 responses with high enough fidelity without severe complication.

See https://github.com/linkerd/linkerd/issues/1847, which is an open issue against Linkerd that seems to be exactly this issue. In general, translating between two versions, even HTTP 1.0 and HTTP 1.1, is a lot of tedious and error-prone work. If we're going to spend that effort, I'd rather first spend it on getting (some subset of) HTTP/1.1 upgraded to HTTP/2 over TLS.

> I'm a little sympathetic to the case where clients or servers only support 1.0, and are so poorly written that they blow up otherwise, but how often do these implementations exist?

I think we have to assume that they are common. There are many sloppy HTTP/1.x implementations that work *just* well enough to support whatever peers they already talk to.

Reply:
> For example, advertising HTTP/1.1 in our request means we support chunked transfer encoding in responses, and we'd have to translate that into a non-chunked response.

In that case, it'd be OK! hyper knows that an HTTP/1.0 response is not allowed to be chunked, and so removes that header and converts to a close-delimited body.

> I think we have to assume that they are common.

Perhaps this one of those things that we can take advantage of the experience of Linkerd. If they've found that it's not that common, but upgrading to 1.1 is better for many people, then we can adjust our assumptions.

Reply:
> For example, advertising HTTP/1.1 in our request means we support chunked transfer encoding in responses, and we'd have to translate that into a non-chunked response.
> 
> In that case, it'd be OK! hyper knows that an HTTP/1.0 response is not allowed to be chunked, and so removes that header and converts to a close-delimited body.

There must be lots of apps that expect a Content-Length to be provided in responses as close-terminated bodies are degenerate behavior. I know I've written some apps that assume Content-Length is in the response.

More generally, there's definitely potential ways to upgrade HTTP/1.0 to HTTP/1.1. However, I'm skeptical about the cost/benefit at the current point in time. It's tedious and time-consuming to consider ever difference between HTTP/1.0 and HTTP/1.1 and how to deal with them. Whereas, AFAICT proxying HTTP/1.0 traffic as HTTP/1.0 works fine and doesn't hurt our primary value proposition, whichI see as being about improving observability, reliabilty, and security without compromising performance. Even more generally, I think HTTP/1.0 source traffic is really a degenerate case to start with, so I don't understand why we want to spend effort on optimizing it. Surely the vast majority of source traffic is going to be HTTP/1.1?

> Perhaps this one of those things that we can take advantage of the experience of Linkerd. If they've found that it's not that common, but upgrading to 1.1 is better for many people, then we can adjust our assumptions.

I agree it's good to look at linkerd's experience. My general impression of linkerd's experience is that it is what motivated us to be highly transparent to start with. Also absence of evidence isn't evidence of absence (sorry for the aphorism; I don't know another way to express it).